cli: TestHalfOnlineLossOfQuorumRecovery failed

cli.TestHalfOnlineLossOfQuorumRecovery [failed](https://tanzanite.cluster.engflow.com/invocations/default/ca69ddd9-a359-4e97-b9d2-7800e5d8a1dd?testReportRun=2&testReportShard=14&testReportAttempt=1#targets-Ly9wa2cvY2xpOmNsaV90ZXN0) on master @ [93ad913106b6f0f6ec98bc2cfa788ff6d8085bd4](https://github.com/cockroachdb/cockroach/commits/93ad913106b6f0f6ec98bc2cfa788ff6d8085bd4):


```
=== RUN   TestHalfOnlineLossOfQuorumRecovery
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestHalfOnlineLossOfQuorumRecovery557329092
    test_log_scope.go:81: use -show-logs to present logs inline
[debug recover make-plan --confirm=y --certs-dir=test_certs --host=127.0.0.1:34717 --plan=/var/lib/engflow/worker/work/3/exec/_tmp/2dcc0a708d6e811d98e64cdd1ab12ba9/TestHalfOnlineLossOfQuorumRecovery916069754/recovery-plan.json]
    debug_recover_loss_of_quorum_test.go:535: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/cli/debug_recover_loss_of_quorum_test.go:535
        	Error:      	"debug recover make-plan --confirm=y --certs-dir=test_certs --host=127.0.0.1:34717 --plan=/var/lib/engflow/worker/work/3/exec/_tmp/2dcc0a708d6e811d98e64cdd1ab12ba9/TestHalfOnlineLossOfQuorumRecovery916069754/recovery-plan.json\nERROR: failed to retrieve replica info from cluster: rpc error: code = Unknown desc = failed retrieving replicas from node n1 during fan-out: recv msg error: grpc: key /Local/RangeID/1/r/RangeGCThreshold does not have /Local/Range prefix [code 2/Unknown]\n" does not contain "- node n1"
        	Test:       	TestHalfOnlineLossOfQuorumRecovery
        	Messages:   	planner didn't provide correct apply instructions
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestHalfOnlineLossOfQuorumRecovery557329092
--- FAIL: TestHalfOnlineLossOfQuorumRecovery (31.38s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=14</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHalfOnlineLossOfQuorumRecovery.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-38861

cli.TestHalfOnlineLossOfQuorumRecovery [failed](https://tanzanite.cluster.engflow.com/invocations/default/f613e5e9-7c6c-497e-9c6f-6b00834f2d46?testReportRun=2&testReportShard=14&testReportAttempt=1#targets-Ly9wa2cvY2xpOmNsaV90ZXN0) on master @ [d146ecff6f687e438706cf63591cafca60cc116d](https://github.com/cockroachdb/cockroach/commits/d146ecff6f687e438706cf63591cafca60cc116d):


```
=== RUN   TestHalfOnlineLossOfQuorumRecovery
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestHalfOnlineLossOfQuorumRecovery3317157959
    test_log_scope.go:81: use -show-logs to present logs inline
[debug recover make-plan --confirm=y --certs-dir=test_certs --host=127.0.0.1:43875 --plan=/var/lib/engflow/worker/work/2/exec/_tmp/2dcc0a708d6e811d98e64cdd1ab12ba9/TestHalfOnlineLossOfQuorumRecovery1639678948/recovery-plan.json]
    debug_recover_loss_of_quorum_test.go:535: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/cli/debug_recover_loss_of_quorum_test.go:535
        	Error:      	"debug recover make-plan --confirm=y --certs-dir=test_certs --host=127.0.0.1:43875 --plan=/var/lib/engflow/worker/work/2/exec/_tmp/2dcc0a708d6e811d98e64cdd1ab12ba9/TestHalfOnlineLossOfQuorumRecovery1639678948/recovery-plan.json\nERROR: failed to retrieve replica info from cluster: rpc error: code = Unknown desc = failed retrieving replicas from node n1 during fan-out: recv msg error: grpc: key /Local/RangeID/1/r/RangeGCThreshold does not have /Local/Range prefix [code 2/Unknown]\n" does not contain "- node n1"
        	Test:       	TestHalfOnlineLossOfQuorumRecovery
        	Messages:   	planner didn't provide correct apply instructions
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestHalfOnlineLossOfQuorumRecovery3317157959
--- FAIL: TestHalfOnlineLossOfQuorumRecovery (46.72s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=14</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHalfOnlineLossOfQuorumRecovery.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

cli.TestHalfOnlineLossOfQuorumRecovery [failed](https://tanzanite.cluster.engflow.com/invocations/default/1ce408f4-c13b-46c4-ae87-8bab761072c7?testReportRun=2&testReportShard=14&testReportAttempt=1#targets-Ly9wa2cvY2xpOmNsaV90ZXN0) on master @ [0ea8b622c5354a891cfb867d8d872a0746ba77b3](https://github.com/cockroachdb/cockroach/commits/0ea8b622c5354a891cfb867d8d872a0746ba77b3):


```
=== RUN   TestHalfOnlineLossOfQuorumRecovery
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestHalfOnlineLossOfQuorumRecovery4224013447
    test_log_scope.go:81: use -show-logs to present logs inline
[debug recover make-plan --confirm=y --certs-dir=test_certs --host=127.0.0.1:37559 --plan=/var/lib/engflow/worker/work/0/exec/_tmp/2dcc0a708d6e811d98e64cdd1ab12ba9/TestHalfOnlineLossOfQuorumRecovery3625772260/recovery-plan.json]
    debug_recover_loss_of_quorum_test.go:535: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/cli/debug_recover_loss_of_quorum_test.go:535
        	Error:      	"debug recover make-plan --confirm=y --certs-dir=test_certs --host=127.0.0.1:37559 --plan=/var/lib/engflow/worker/work/0/exec/_tmp/2dcc0a708d6e811d98e64cdd1ab12ba9/TestHalfOnlineLossOfQuorumRecovery3625772260/recovery-plan.json\nERROR: failed to retrieve replica info from cluster: rpc error: code = Unknown desc = failed retrieving replicas from node n1 during fan-out: recv msg error: grpc: key /Local/RangeID/1/r/RangeGCThreshold does not have /Local/Range prefix [code 2/Unknown]\n" does not contain "- node n1"
        	Test:       	TestHalfOnlineLossOfQuorumRecovery
        	Messages:   	planner didn't provide correct apply instructions
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestHalfOnlineLossOfQuorumRecovery4224013447
--- FAIL: TestHalfOnlineLossOfQuorumRecovery (32.06s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=14</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHalfOnlineLossOfQuorumRecovery.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

cli.TestHalfOnlineLossOfQuorumRecovery [failed](https://tanzanite.cluster.engflow.com/invocations/default/a3b1bed2-d780-4029-bce2-33fd9fe44f3d?testReportRun=3&testReportShard=14&testReportAttempt=1#targets-Ly9wa2cvY2xpOmNsaV90ZXN0) on master @ [e5f5597b1c1570fff03ee6dfd4a1e8be65c0e008](https://github.com/cockroachdb/cockroach/commits/e5f5597b1c1570fff03ee6dfd4a1e8be65c0e008):


```
=== RUN   TestHalfOnlineLossOfQuorumRecovery
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestHalfOnlineLossOfQuorumRecovery1927781118
    test_log_scope.go:81: use -show-logs to present logs inline
[debug recover make-plan --confirm=y --certs-dir=test_certs --host=127.0.0.1:40383 --plan=/var/lib/engflow/worker/work/2/exec/_tmp/627c07b75a1b49827ad9adf3e9f30827/TestHalfOnlineLossOfQuorumRecovery4013921476/recovery-plan.json]
    debug_recover_loss_of_quorum_test.go:535: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/cli/debug_recover_loss_of_quorum_test.go:535
        	Error:      	"debug recover make-plan --confirm=y --certs-dir=test_certs --host=127.0.0.1:40383 --plan=/var/lib/engflow/worker/work/2/exec/_tmp/627c07b75a1b49827ad9adf3e9f30827/TestHalfOnlineLossOfQuorumRecovery4013921476/recovery-plan.json\nERROR: failed to retrieve replica info from cluster: rpc error: code = Unknown desc = failed retrieving replicas from node n1 during fan-out: recv msg error: grpc: key /Local/RangeID/1/r/RangeGCThreshold does not have /Local/Range prefix [code 2/Unknown]\n" does not contain "- node n1"
        	Test:       	TestHalfOnlineLossOfQuorumRecovery
        	Messages:   	planner didn't provide correct apply instructions
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestHalfOnlineLossOfQuorumRecovery1927781118
--- FAIL: TestHalfOnlineLossOfQuorumRecovery (32.19s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=14</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHalfOnlineLossOfQuorumRecovery.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

This is very strange. With some debugging info added:
```
panic: key: /Local/RangeID/1/r/RangeGCThreshold  "\x01i\x89rlgc-"  Last SeekGE() key: empty timestamp case /Local/Range/Table/Max/RangeDescriptor/1717532590.198522999,2147483647  "\x01k\x12\xfa\x00\x01rdsc"  NextKey() calls since: 0
```

We have an `MVCCIterator` which we `SeekGE` to `\x01k\x12\xfa\x00\x01rdsc`. The iterator returns key "\x01i\x89rlgc-" which is smaller. This is a problem in `MVCCIterator` or inside Pebble.

--- a/pkg/kv/kvserver/kvstorage/init.go
+++ b/pkg/kv/kvserver/kvstorage/init.go
@@ -271,6 +271,7 @@ func IterateRangeDescriptorsFromDisk(
 	lastReportTime := timeutil.Now()
 
 	iter.SeekGE(storage.MVCCKey{Key: keys.LocalRangePrefix})
+	var keyBuf []byte
 	for {
 		if valid, err := iter.Valid(); err != nil {
 			return err
@@ -330,7 +331,9 @@ func IterateRangeDescriptorsFromDisk(
 				return errors.AssertionFailedf("range key has intent with no timestamp")
 			}
 			// Seek to the latest value below the intent timestamp.
-			iter.SeekGE(storage.MVCCKey{Key: key.Key, Timestamp: metaTS.Prev()})
+			// We cannot pass to SeekGE a key that was returned to us from the iterator; make a copy.
+			keyBuf = append(keyBuf[:0], key.Key...)
+			iter.SeekGE(storage.MVCCKey{Key: keyBuf, Timestamp: metaTS.Prev()})
 			continue
 		}
 

